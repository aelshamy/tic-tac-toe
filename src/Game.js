import React from "react";
import Board from "./Board";

class Game extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      values: Array(9).fill(null),
      playerIsNext: true
    };
  }

  handleClick = idx => {
    const values = [...this.state.values];
    if (values[idx]) {
      return;
    }
    values[idx] = this.state.playerIsNext ? "X" : "O";
    this.setState({
      playerIsNext: !this.state.playerIsNext,
      values
    });
  };

  render() {
    return (
      <div className="game">
        <Board values={this.state.values} onClick={this.handleClick} />
      </div>
    );
  }
}

export default Game;
