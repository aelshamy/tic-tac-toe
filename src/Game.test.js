import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import { create } from "react-test-renderer";
import Game from "./Game";

describe("<Game/>", () => {
  let game = Game;
  beforeEach(() => {
    game = new Game();
  });

  test("it will match the snapshot", () => {
    const c = create(<Game />);
    expect(toJson(c)).toMatchSnapshot();
  });

  test("should intialize with the default values", () => {
    expect(game.state).not.toBe(null);
    expect(game.state.values.length).toEqual(9);
    expect(game.state.values[0]).toEqual(null);
    expect(game.state.playerIsNext).toEqual(true);
  });

  describe("calculateWinner()", () => {
    test("should return null if the provided values is not one of the wining squares", () => {
      let result = game.calculateWinner([]);
      expect(result).toBe(null);
      result = game.calculateWinner(["X", "X", "O"]);
      expect(result).toBe(null);
    });

    test("should return the winner if the provided value one of the wining squares", () => {
      const result = game.calculateWinner(["X", "X", "X"]);
      expect(result).toBe("X");
    });
  });

  describe("handleClick()", () => {
    test("should return if there is a winner", () => {
      spyOn(game, "calculateWinner").and.returnValue("X");
      spyOn(game, "setState");
      game.handleClick(1);
      expect(game.setState).not.toHaveBeenCalled();
    });

    test("should return if user clicked on existed square", () => {
      game.state.values = [null, "X", null];
      spyOn(game, "setState");
      const result = game.handleClick(1);
      expect(game.setState).not.toHaveBeenCalled();
    });

    test("should set the correct values to the state", () => {
      spyOn(game, "setState");
      const index = 1;
      const expectedValues = [...game.state.values];
      expectedValues[index] = "X";
      game.handleClick(index);
      expect(game.setState).toHaveBeenCalledWith({
        playerIsNext: !game.state.playerIsNext,
        values: expectedValues
      });
    });
  });

  describe("resetGame()", () => {
    test("should rest game to default", () => {
      spyOn(game, "setState");
      game.resetGame();
      expect(game.setState).toHaveBeenCalledWith({
        values: Array(9).fill(null),
        playerIsNext: true
      });
    });
  });

  describe("render()", () => {
    test("should render game correctly", () => {
      const component = shallow(<Game />);
      expect(component).toMatchSnapshot();
      expect(component.find(".game").exists()).toBe(true);
      expect(component.find("Board").exists()).toBe(true);
      expect(component.find("result").exists()).toBe(false);
    });

    test("should render game with won result", () => {
      const component = shallow(<Game />);
      component.setState({
        values: ["X", "X", "X"],
        playerIsNext: false
      });

      // console.log(component.state());
      expect(component.find("Board").exists()).toBe(false);
      expect(component.find(".result").exists()).toBe(true);
      expect(
        component
          .find(".result")
          .find(".winner")
          .exists()
      ).toBe(true);
      expect(
        component
          .find(".result")
          .find(".loser")
          .exists()
      ).toBe(false);
      expect(
        component
          .find(".result")
          .find("button")
          .exists()
      ).toBe(true);
    });
    test("should render game with won result", () => {
      const component = shallow(<Game />);
      component.setState({
        values: ["O", "O", "O"],
        playerIsNext: true
      });

      expect(component.find("Board").exists()).toBe(false);
      expect(component.find(".result").exists()).toBe(true);
      expect(
        component
          .find(".result")
          .find(".winner")
          .exists()
      ).toBe(false);
      expect(
        component
          .find(".result")
          .find(".loser")
          .exists()
      ).toBe(true);
      expect(
        component
          .find(".result")
          .find("button")
          .exists()
      ).toBe(true);
    });
  });
});
