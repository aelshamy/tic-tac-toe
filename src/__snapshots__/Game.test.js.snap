// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game/> it will match the snapshot 1`] = `undefined`;

exports[`<Game/> render() should render correctly 1`] = `
<div
  className="game"
>
  <Board
    onClick={[Function]}
    values={
      Array [
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
      ]
    }
  />
</div>
`;
